pkg/envoy/admin/admin.go

```diff
@@ -0,0 +1,143 @@
+package admin
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"crypto/tls"
+	"encoding/hex"
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/kumahq/kuma/pkg/core"
+
+	"github.com/kumahq/kuma/pkg/core/resources/apis/system"
+
+	"github.com/pkg/errors"
+
+	kuma_cp "github.com/kumahq/kuma/pkg/config/app/kuma-cp"
+	mesh_core "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	"github.com/kumahq/kuma/pkg/core/resources/manager"
+	core_store "github.com/kumahq/kuma/pkg/core/resources/store"
+	"github.com/kumahq/kuma/pkg/tokens/builtin/issuer"
+)
+
+const envoyAdminClientTokenPrefix = "envoy-admin-client-token"
+
+var log = core.Log.WithName("envoy-admin")
+
+type EnvoyAdminClient interface {
+	GenerateAPIToken(dataplane *mesh_core.DataplaneResource) (string, error)
+	PostQuit(dataplane *mesh_core.DataplaneResource) error
+}
+
+type envoyAdminClient struct {
+	rm         manager.ResourceManager
+	cfg        kuma_cp.Config
+	httpClient *http.Client
+	scheme     string
+}
+
+func NewEnvoyAdminClient(rm manager.ResourceManager, cfg kuma_cp.Config) EnvoyAdminClient {
+	return &envoyAdminClient{
+		rm:  rm,
+		cfg: cfg,
+		httpClient: &http.Client{
+			Transport: &http.Transport{
+				Dial: (&net.Dialer{
+					Timeout: 3 * time.Second,
+				}).Dial,
+				TLSHandshakeTimeout: 3 * time.Second,
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: true,
+				},
+			},
+			Timeout: 5 * time.Second,
+		},
+		scheme: "https",
+	}
+}
+
+const (
+	quitquitquit = "quitquitquit"
+)
+
+func (a *envoyAdminClient) GenerateAPIToken(dataplane *mesh_core.DataplaneResource) (string, error) {
+	mesh := dataplane.Meta.GetMesh()
+	key, err := a.getSigningKeyString(mesh)
+	if err != nil {
+		return "", err
+	}
+
+	mac := hmac.New(sha256.New, []byte(key))
+	_, err = mac.Write([]byte(dataplane.Meta.GetName()))
+	if err != nil {
+		return "", err
+	}
+
+	return hex.EncodeToString(mac.Sum(nil)), nil
+}
+
+func (a *envoyAdminClient) getSigningKeyString(mesh string) (string, error) {
+	signingKey := system.NewSecretResource()
+	key := issuer.SigningKeyResourceKey(envoyAdminClientTokenPrefix, mesh)
+	err := a.rm.Get(context.Background(), signingKey, core_store.GetBy(key))
+	if err == nil {
+		return signingKey.Spec.GetData().String(), nil
+	}
+	if !core_store.IsResourceNotFound(err) {
+		return "", errors.Wrap(err, "could not retrieve a resource")
+	}
+
+	// Key not found, create it
+	signingKey, err = issuer.CreateSigningKey()
+	if err != nil {
+		return "", errors.Wrap(err, "could not create a signing key")
+	}
+	if err := a.rm.Create(context.Background(), signingKey, core_store.CreateBy(key)); err != nil {
+		return "", errors.Wrap(err, "could not create a resource")
+	}
```

Comment:
How does it work with Multizone?

I think it should be created on Global just like the signing zone for dp tokens. If we complete our goal of getting rid of DB dependency on remotes, this value won't be consistent between instances of the CP, whereas if it's generated on Global, it will be.